fsharp-backend/tests/Tests/Serialization.Tests.fs

```diff
@@ -0,0 +1,999 @@
+  //   change, etc. You may want to save both formats and check that both are
+  //   deserializable.
+  // - If there's a change to the sample value, generate a new file.
+  //   Commit it after verifying that the format is appropriate.
+
+
+  module SampleData =
+    open Values
+
+    // Stores a list of sample outputs for `$"{serializerName}-{typeName}"
+    let data = Dictionary.T<string, List<string * string>>()
+
+    let keyFor (typeName : string) (serializerName : string) : string =
+      $"{serializerName}-{typeName}"
+
+    let fileNameFor
+      (typeName : string)
+      (serializerName : string)
+      (dataName : string)
+      : string =
+      let original = $"{serializerName}-{typeName}-{dataName}"
+      let replaced = Regex.Replace(original, "[^-_a-zA-Z0-9]", "-")
+      Regex.Replace(replaced, "[-]+", "-") + ".json"
+
+    let get
+      (typeName : string)
+      (serializerName : string)
+      (reason : string)
+      : List<string * string> =
+      data
+      |> Dictionary.get (keyFor typeName serializerName)
+      |> Exception.unwrapOptionInternal
+           "testCases should exist for allowed serializable type"
+           [ "typeName", typeName :> obj
+             "serializer", serializerName
+             "reason", reason ]
+
+
+    let generate<'v>
+      (serializerName : string)
+      (f : 'v -> string)
+      (name : string)
+      (v : 'v)
+      =
+      let typeName = string typeof<'v>
+      let key = keyFor typeName serializerName
+      let currentValue = Dictionary.get key data |> Option.unwrap []
+      let newValue = (name, f v) :: currentValue
+      data[key] <- newValue
+
+    let v<'t when 't : equality> (dataName : string) (data : 't) =
+      // let roundtripped =
+      //   data |> Json.Vanilla.serialize |> Json.Vanilla.deserialize<'t>
+      // if data <> roundtripped then
+      //   Exception.raiseInternal
+      //     $"bad vanilla roundtrip for {dataName}"
+      //     [ "actual", string roundtripped; "expected", string data ]
+
+      // Use prettySerialize even though we use serialize in practice, as the
+      // non-pretty version is too hard to read and compare.
+      generate "vanilla" Json.Vanilla.prettySerialize dataName data
+
+    let oc<'t when 't : equality> (dataName : string) (data : 't) =
+      // let roundtripped =
+      //   data |> Json.Vanilla.serialize |> Json.Vanilla.deserialize<'t>
+      // if data <> roundtripped then
+      //   Exception.raiseInternal
+      //     $"bad ocaml roundtrip for {dataName}"
+      //     [ "actual", roundtripped; "expected", string data ]
+      // Use prettySerialize even though we use serialize in practice, as the
+      // non-pretty version is too hard to read and compare.
+      generate "ocaml" Json.OCamlCompatible.prettySerialize dataName data
+
+    let both<'t when 't : equality> (dataName) (data : 't) =
+      v<'t> dataName data
+      oc<'t> dataName data
+
+    let generateTestData () : unit =
```

Comment:
I'm not sure why this can't be a function that _returns_ a `Dictionary.T<string, List<string * string>>()`, rather than the current indirection of setting it up and having the various generation functions make insertions into it.

Reply:
I just spent some time trying to set this up, and it came out worse (lists of things, indentation troubles). So I left it.